File: CS/Models/XpoModelBinder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using DevExpress.Web.Mvc;
using System.Web.Mvc;
using DevExpress.Xpo.Metadata;
using DevExpress.Xpo;
using System.ComponentModel;

public class XpoModelBinder : DevExpressEditorsBinder {
    protected override object CreateModel(ControllerContext controllerContext, ModelBindingContext bindingContext, Type modelType) {
        IXpoController xpoController = controllerContext.Controller as IXpoController;

        if (xpoController == null)
            throw new InvalidOperationException("The controller does not support IXpoController interface");

        XPClassInfo classInfo = xpoController.XpoSession.GetClassInfo(modelType);
        ValueProviderResult result = bindingContext.ValueProvider.GetValue(classInfo.KeyProperty.Name);

        return result == null ? classInfo.CreateNewObject(xpoController.XpoSession) :
            xpoController.XpoSession.GetObjectByKey(classInfo, result.ConvertTo(classInfo.KeyProperty.MemberType));
    }
}

File: VB/Models/XpoModelBinder.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports DevExpress.Web.Mvc
Imports System.Web.Mvc
Imports DevExpress.Xpo.Metadata
Imports DevExpress.Xpo
Imports System.ComponentModel

Public Class XpoModelBinder
	Inherits DevExpressEditorsBinder
	Protected Overrides Function CreateModel(ByVal controllerContext As ControllerContext, ByVal bindingContext As ModelBindingContext, ByVal modelType As Type) As Object
		Dim xpoController As IXpoController = TryCast(controllerContext.Controller, IXpoController)

		If xpoController Is Nothing Then
			Throw New InvalidOperationException("The controller does not support IXpoController interface")
		End If

		Dim classInfo As XPClassInfo = xpoController.XpoSession.GetClassInfo(modelType)
		Dim result As ValueProviderResult = bindingContext.ValueProvider.GetValue(classInfo.KeyProperty.Name)

		Return If(result Is Nothing, classInfo.CreateNewObject(xpoController.XpoSession), xpoController.XpoSession.GetObjectByKey(classInfo, result.ConvertTo(classInfo.KeyProperty.MemberType)))
	End Function
End Class